### PR TITLE
Add `ArticlePlacement` class

### PR DESCRIPTION
**Pull request by JUNYUANGO**
Hi, I am considering to have a ScrapedArticle class instead of using Article directly. Please review if this is reasonable. I already tested this code with ones in the PR inside the poprox-storage repo. They works.

### CODE REVIEW

**Comment by karlhigley** [comment]
I think you're on the right general track here—I don't think we should use `Article` and we probably do want to have a new class in this library to represent a new concept we're adding. I'm not sure that new concept is scraped articles though, since this looks like it has a lot of overlap with `Article`.

If we dropped the parts that `Article` already has, it looks like we'd end up with `section` and `level`, but what would we call that? `ArticlePosition`?

**Comment by JUNYUANGO** [comment]
> I think you're on the right general track here—I don't think we should use `Article` and we probably do want to have a new class in this library to represent a new concept we're adding. I'm not sure that new concept is scraped articles though, since this looks like it has a lot of overlap with `Article`.
> 
> If we dropped the parts that `Article` already has, it looks like we'd end up with `section` and `level`, but what would we call that? `ArticlePosition`? `ArticlePlacement`?
> 
> ```python
> class ArticlePlacement:
>     article_id: UUID
>     section: str | None = None
>     level: str | None = None
> ```

File: src/poprox_concepts/domain/article.py [hunk 1]
```diff
@@ -39,3 +39,13 @@ class ArticleSet(BaseModel):
...
+
+
+class ScrapedArticle(BaseModel):
+    title: str
+    description: str | None = None
+    url: str
+    section: str | None = None
+    level: str | None = None
+    image_url: str | None = None
+    created_at: datetime = datetime.now(timezone.utc)
```
Comment (kluver) [inline src/poprox_concepts/domain/article.py:51]: We don't always include created_at in our model objects -- two reasons. First we don't really want this auto-assigning to when the model object was created like it does here -- the source of truth for that should probably be the database. But secondly, we just rarely need that information in python so we add it only when we know we need it.

**Comment by kluver** [comment]
> Last time I met with Sophia, Ishan, and Edward, we discussed about adding more properties? for Article, but I felt that it might change a lot here, since we didn't want to mess up the articles from AP News with the scraped ones, I guess?

The way I might look at it is not that the current proposal is incorrect. I think there is a world in which we press "merge". And I think there is a world in which we just update Article. The place I think we're at is picking the best of several workable solutions.

On that front. I agree with @JUNYUANGO that updating the article model as-exists is probably not the best option. That "concept" is already loaded with a lot of different meanings and adding 1 more might break it.

I like the idea of an `ArticlePlacement` object. But I have two thoughts:

- @JUNYUANGO, were the description and title fields going to store _API data_ or _scraped data_? Because if the semantic meaning is "this is the exact content we saw when we scrapped" then that isn't necessarily a duplicate of what's already in the database. There's room to ask "do we care about this difference, and will we use it?" but it's not the same.
- @karlhigley do we want to look at this new object as a better way for recommenders to return their desired article placement in a newsletter. It's theoretically wasteful that the recommender has to send entire article elements back to us, but right now I don't know if we have a better solution.

**Comment by karlhigley** [comment]
> @karlhigley do we want to look at this new object as a better way for recommenders to return their desired article placement in a newsletter. It's theoretically wasteful that the recommender has to send entire article elements back to us, but right now I don't know if we have a better solution.

@kluver I was intending this just to capture the placement of the article on AP's website. It's not really clear to me what a reasonably general way to capture this information would be, so I'm trying on some different conceptual frames.

**Comment by kluver** [comment]
> > @karlhigley do we want to look at this new object as a better way for recommenders to return their desired article placement in a newsletter. It's theoretically wasteful that the recommender has to send entire article elements back to us, but right now I don't know if we have a better solution.
> 
> @kluver I was intending this just to capture the placement of the article on AP's website. It's not really clear to me what a reasonably general way to capture this information would be, so I'm trying on some different conceptual frames.

Sure -- let's get this captured well and then think about other topics later then.

@JUNYUANGO What are you thinking here? Any of these designs feeling particularly good/bad?

**Comment by JUNYUANGO** [comment]
> * @JUNYUANGO, were the description and title fields going to store _API data_ or _scraped data_? Because if the semantic meaning is "this is the exact content we saw when we scrapped" then that isn't necessarily a duplicate of what's already in the database. There's room to ask "do we care about this difference, and will we use it?" but it's not the same.
> * @JUNYUANGO What are you thinking here? Any of these designs feeling particularly good/bad?

1. Both the description and title fields will store the _scraped data_, since we already have _API data_, so that we can have a space to find out if there is a difference between the _API data_ and _scraped data_ (or if AP News is doing a A/B test?). 
2. Based on what we have discussed above, I would prefer to have an ArticlePlacement (? not exactly understand what that means) to keep the most important (all would be used) information: title, description, section, level, and image_url (for Ishan's matching and classification work). 

I feel that this could be an additional layer?feature? of the system, without having too much influence on the current system. If this looks good for you all as well, I will work on the code to match that.

**Comment by karlhigley** [comment]
> ArticlePlacement (? not exactly understand what that means)

The distinction I'm trying to make is that there are actually three things in play:
* Raw data from AP's feed API
* Raw data scraped from AP's website
* The concept of an `Article` in our domain/data model, which isn't the same as either of the above

In the same way that we don't need everything that comes from the API stored in our database, we also don't need everything that was scraped—especially when that will lead to a un-normalized database schema. If we store the article title in a new table, we're duplicating data that already exists in another table and creating multiple sources of truth for the article attributes. So, to avoid that, I'm trying to figure out what _new_ pieces of information we need to store and what to call them.

It seems like the new information here isn't the article title or URL, but instead information about where articles where and how displayed on the website. `ArticlePlacement` is me trying to find a way to turn "where the article was displayed and how large it was" into a noun that makes sense as a database table name and a domain concept.

If we went in that direction, we'd end up with a database table that links to the `articles` table with a foreign key:
```
                                  Table "public.article_placements"
      Column      |            Type             | Collation | Nullable |      Default      
------------------+-----------------------------+-----------+----------+-------------------
 article_id       | uuid                        |           | not null |
 section          | character varying           |           | not null | 
 level            | character varying           |           |          | 
 created_at       | timestamp without time zone |           | not null | now()
 Foreign-key constraints:
    "fk_article_placements_article_id" FOREIGN KEY (article_id) REFERENCES articles(article_id)

```
And we'd create records in that table with the _output_ of the linking process instead of the input.

**Comment by JUNYUANGO** [comment]
> If we went in that direction, we'd end up with a database table that links to the `articles` table with a foreign key:
> 
> ```
>                                   Table "public.article_placements"
>       Column      |            Type             | Collation | Nullable |      Default      
> ------------------+-----------------------------+-----------+----------+-------------------
>  article_id       | uuid                        |           | not null |
>  section          | character varying           |           | not null | 
>  level            | character varying           |           |          | 
>  created_at       | timestamp without time zone |           | not null | now()
>  Foreign-key constraints:
>     "fk_article_placements_article_id" FOREIGN KEY (article_id) REFERENCES articles(article_id)

This looks good and clear to me. I would ask if Ishan needs image_url for his work. I will push new codes for this table later.

**Review by karlhigley — APPROVED** [review]

**Pull request merged**